Closes #6638: Handle Properly picture tag

Miraeld opened the pull request:
# Description

Fixes #6638 #6660 #6669

## Documentation

### User documentation
First of all, this PR brings a fix for the LCP candidates when it's a `<picture>` tag detected. Before, we would also detect the sub-tag `<img>` and both would compete each other leading to errors. 
Then, it also brings another way to process `<picture>` tag when there is a `type` attribute in order to preload the first `<source>` image. 

### Technical documentation

## Type of change
- [x] Enhancement (non-breaking change which improves an existing functionality).

## New dependencies

## Risks

# Checklists

## Feature validation

- [x] I validated all the Acceptance Criteria. If possible, provide sreenshots or videos.
- [x] I triggered all changed lines of code at least once without new errors/warnings/notices.
- [x] I implemented built-in tests to cover the new/changed code.

## Documentation

- [x] I prepared the user documentation for the feature/enhancement and shared it in the PR or the GitHub issue.
- [x] The user documentation covers new/changed entry points (endpoints, WP hooks, configuration files, ...).
- [x] I prepared the technical documentation if needed, and shared it in the PR or the GitHub issue.

## Code style
- [x] I wrote self-explanatory code about what it does.
- [x] I wrote comments to explain why it does it.
- [x] I named variables and functions explicitely.
- [x] I protected entry points against unexpected inputs.
- [x] I did not introduce unecessary complexity.
- [x] I listed the introduced external dependencies explicitely on the PR.
- [x] I validated the repo-specific guidelines from CONTRIBUTING.md.

## Observability
- [x] I handled errors when needed.
- [x] I wrote user-facing messages that are understandable and provide actionable feedbacks.
- [x] I prepared ways to observe the implemented system (logs, data, etc.).

## Risks
- [x] I explicitely mentioned performance risks in the PR.
- [x] I explicitely mentioned security risks in the PR.

Miraeld committed:
Fix subtag <img> of <picture> being picked-up as lcp candidates

Miraeld committed:
modify process of <picture> tag to comply AC

Miraeld committed:
Add tests

Miraeld committed:
Fix conflicts

Miraeld committed:
Fix linter

remyperona commented on inc/Engine/Media/AboveTheFold/Frontend/Controller.php:
```diff
@@ -437,11 +437,24 @@ private function generate_source_tags( $lcp, $start_tag, $end_tag ) {
+				// Generate the link tag.
+				$tag .= $start_tag . 'href="' . $source->srcset . '" ' . ( $sizes ?? '' ) . $end_tag;
+				break;
+			} else {
```
the usage of `else` here is not necessary if you `break` just before I believe

Miraeld replied:
Done

Miraeld committed:
comply feedback

Miraeld committed:
Closes #6660: Handle media properly in <picture>

Miraeld committed:
Closes #6669:
Fix the empty href & duplicated `min-width` with picture tag

remyperona commented:
Need to lint the code a bit

Miraeld committed:
fix linter

Miraeld commented:
My bad, didn't see that failed CI. 
Done :)

Mai-Saad commented:
Test plan:
- run e2e tests => no regression
- check the fetch priority is correct + images excluded from LL for  mobile and desktop:
Different media (AC https://github.com/wp-media/wp-rocket/issues/6530#issue-2218249252) ==> https://e2e.rocketlabsqa.ovh/lcp_picture_template2/ :heavy_check_mark:  , 
https://e2e.rocketlabsqa.ovh/lcp_picture_template/ => having min-width only is agreed as edge case now https://wp-media.slack.com/archives/CUT7FLHF1/p1716990574011419?thread_ts=1716965466.492079&cid=CUT7FLHF1, 
https://e2e.rocketlabsqa.ovh/lcp_picture_relative/
Different type (only 1st type preloaded in 1st 2 templates, AC https://github.com/wp-media/wp-rocket/issues/6638#issue-2302642471) ==> https://e2e.rocketlabsqa.ovh/lcp_picture_type/, https://e2e.rocketlabsqa.ovh/lcp_picture_responsive_type/, https://e2e.rocketlabsqa.ovh/lcp_picture_media_type_mixed/

Miraeld committed:
Fix last AC

Miraeld committed:
Merge branch 'develop' into enhancement/6638-handle-picture-tag-in-lcp-preload

Miraeld commented:
Good morning @Mai-Saad & @piotrbak ,

Latest commits are adding the following: 
- The newest AC (https://github.com/wp-media/wp-rocket/issues/6638#issuecomment-2139333126), 
- Fix tests, each AC are included in integration tests, however something was wrong with them, and they were not running. Now it's running and everything is passing. 

Also, e2e is passing on my side, everything ok.

Miraeld committed:
Merge branch 'develop' into enhancement/6638-handle-picture-tag-in-lcp-preload

Miraeld committed:
beacon minified version update

Miraeld commented:
@Mai-Saad if it can help for your tests, 
Note that every cases are included in the integration test for the front-end, it will check that we are processign and adding the preload to images as per requested by ACs. 
As integration tests aren't running Javascript, it does not test the `lcp-beacon.js` part.

Mai-Saad commented:
@Miraeld Thanks for the update. Please find exploratory test notes below  after merging dev to PR:
1- As per 2nd [AC](https://github.com/wp-media/wp-rocket/issues/6638#issuecomment-2139333126), The following markup 
```
	<picture decoding="async">
    <source type="image/avif" srcset="https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1024x576.jpg.avif">
    <source type="image/webp" srcset="https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1024x576.jpg.webp">
    <img decoding="async" src="https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1024x576.jpg">
</picture>
```
shall be preloaded as 
`<link rel="preload" as="image" href="https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1024x576.jpg.avif" fetchpriority="high">`  ==> currently it is `<link rel="preload" as="image" imagesrcset="https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1024x576.jpg.avif"  fetchpriority="high">`

note: href shall be used,  not imgsrcset https://e2e.rocketlabsqa.ovh/lcp_picture_type/

2- As per 1st AC 
  ```
        <picture  decoding="async" class="wp-image-8418">
    <source type="image/avif" srcset="https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1024x576.jpg.avif 1024w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-300x169.jpg.avif 300w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-768x432.jpg.avif 768w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1536x864.jpg.avif 1536w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1200x675.jpg.avif 1200w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-600x338.jpg.avif 600w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia.jpg.avif 1920w" sizes="(max-width: 1024px) 100vw, 1024px">
    <source type="image/webp" srcset="https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1024x576.jpg.webp 1024w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-300x169.jpg.webp 300w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-768x432.jpg.webp 768w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1536x864.jpg.webp 1536w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1200x675.jpg.webp 1200w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-600x338.jpg.webp 600w" sizes="(max-width: 1024px) 100vw, 1024px">
    <img  decoding="async" width="1024" height="576" src="https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1024x576.jpg" alt="" srcset="https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1024x576.jpg 1024w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-300x169.jpg 300w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-768x432.jpg 768w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1536x864.jpg 1536w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1200x675.jpg 1200w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-600x338.jpg 600w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia.jpg 1920w" sizes="(max-width: 1024px) 100vw, 1024px">
</picture>
```

shall be preloaded as
`<link rel="preload" as="image" imagesrcset="https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1024x576.jpg.avif 1024w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-300x169.jpg.avif 300w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-768x432.jpg.avif 768w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1536x864.jpg.avif 1536w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1200x675.jpg.avif 1200w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-600x338.jpg.avif 600w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia.jpg.avif 1920w" imagesizes="(max-width: 1024px) 100vw, 1024px" fetchpriority="high">` 
==> it is now `<link rel="preload" as="image" imagesrcset="https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-â€”-kopia-1024x576.jpg.avif 1024w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-â€”-kopia-300x169.jpg.avif 300w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-â€”-kopia-768x432.jpg.avif 768w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-â€”-kopia-1536x864.jpg.avif 1536w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-â€”-kopia-1200x675.jpg.avif 1200w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-â€”-kopia-600x338.jpg.avif 600w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-â€”-kopia.jpg.avif 1920w" imagesizes="(max-width: 1024px) 100vw, 1024px" fetchpriority="high">`

Note `-â€”` at the end of URL ===> the 1st visit after clear cache (without timestamp) is good but with 2nd refresh and having timestamp (it's corrupted) , it happens too with picture with single image after 2nd refresh causing lcp 404 URL  https://e2e.rocketlabsqa.ovh/lcp_picture_type/ , https://e2e.rocketlabsqa.ovh/lcp_picture_responsive_type/

3- As per https://wp-media.slack.com/archives/CUKB44GNN/p1717398796565639?thread_ts=1711698134.116649&cid=CUKB44GNN, we shall add `fetchpriority='high'` for the `<img` tag inside picture (we do that now for media picture but not for type one) 
Note: if `<img` tag already has `fetchpriority` then we won't add anything there

Miraeld committed:
Fix feedback 1

Miraeld commented:
Thanks @Mai-Saad for your feedback

1. Fixed with latest commit.

2. I'm a bit embarrassed, I can't reproduce it properly. I see the `-â€` in the db, but when I load the page, everything is correct. The `-â€` is due to a special character, which is `—` and is different from `-`. I'm pinging @Khadreal as he did a PR (https://github.com/wp-media/wp-rocket/pull/6653) related to the special characters. Maybe he could confirm if he can reproduce and if it's fixing it. 

3. It's related to special characters again, so @Khadreal , again, can you confirm you handle this situation. From my investigation: 
When we have this html:
```<picture decoding="async">
    <source type="image/avif" srcset="https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1024x576.jpg.avif">
    <source type="image/webp" srcset="https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1024x576.jpg.webp">
    <img decoding="async" src="https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1024x576.jpg">
</picture>
```

When we save the LCP in the DB, we save the following:
```{
  "type": "picture",
  "src": "https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img--kopia-1024x576.jpg",
  "sources": [
    {
      "srcset": "https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1024x576.jpg.avif",
      "media": "",
      "type": "image/avif",
      "sizes": ""
    },
    {
      "srcset": "https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1024x576.jpg.webp",
      "media": "",
      "type": "image/webp",
      "sizes": ""
    }
  ]
}
```

From this we see a difference in the JSON between `sources` and `src`.
In fact, in `sources`, the name of the image is correct, `img-—-kopia-1024x576`
But in `src`  we get `img--kopia`. It's missing a special character.

So while I was trying to understand, I noticed the following:
The beacon script send this `src`:
`"src":"https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-%E2%80%94-kopia-1024x576.jpg"`
while for the srcset, it's sending the special char as it is `—`

The issue is that, [here](https://github.com/wp-media/wp-rocket/blob/985a76bfcf3d4df0a1dd325b208ed0dd4d071246/inc/Engine/Media/AboveTheFold/AJAX/Controller.php#L53) we do a `sanitize_text_field` which deletes the `%E2%80%94`. And therefore, makes the URL wrong in the db. And this may be fix with https://github.com/wp-media/wp-rocket/pull/6653

Khadreal commented:
Sure @Miraeld, I'll check this and revert. I think what we have should be able to handle this but one can't be too sure

Khadreal commented:
@Miraeld the [PR](https://github.com/wp-media/wp-rocket/pull/6653) handles the special character, with html as below, you will get this preload url with the href url encoded
`<link rel="preload" as="image" href="https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-%E2%80%94-kopia-1024x576.jpg" fetchpriority="high">`
```
<picture>
<source type="image/avif" srcset="https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1024x576.jpg.avif">
    <source type="image/webp" srcset="https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1024x576.jpg.webp">
    <img decoding="async" src="https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1024x576.jpg">
</picture>
```

remyperona committed:
Merge branch 'develop' into enhancement/6638-handle-picture-tag-in-lcp-preload

Miraeld commented:
@Mai-Saad is the encoded URL (Which is valid), is ok or not ?

MathieuLamiot commented:
@Miraeld The PR you are referring to, saying that the test results depend as well on has been merged in develop. You can update your branch (already done) and repeat the test to confirm it is now working completely as expected.

Mai-Saad commented:
@Miraeld Thanks for the update. Here is the latest state (WIP)
1- I can see a problem in 1 but in a different form
this markup:
```
	<picture class="picture_test2">
    <source type="image/avif" srcset ="/wp-content/rocket-test-data/images/lcp/marchlarge1-scaled.jpg.avif">
    <source type="image/webp" srcset="/wp-content/rocket-test-data/images/lcp/marchlarge1-scaled.jpg.webp">
    <img src="/wp-content/rocket-test-data/images/lcp/marchlarge1-scaled.jpg">
</picture>
```
is preloaded as
`<link rel="preload" as="image" href="[/wp-content/rocket-test-data/images/lcp/marchlarge1-scaled.jpg.avif](https://e2e.rocketlabsqa.ovh/wp-content/rocket-test-data/images/lcp/marchlarge1-scaled.jpg.avif)" media=""  fetchpriority="high"><link rel="preload" as="image" href="[/wp-content/rocket-test-data/images/lcp/marchlarge1-scaled.jpg.webp](https://e2e.rocketlabsqa.ovh/wp-content/rocket-test-data/images/lcp/marchlarge1-scaled.jpg.webp)" media=""  fetchpriority="high">`

it should be 

`<link rel="preload" as="image" href="/wp-content/rocket-test-data/images/lcp/marchlarge1-scaled.jpg.avif" fetchpriority="high">`

2,3- still there with the 2nd refresh of page after cache clear (so the page has cache timestamp in signature)

wordpressfan commented:
I tested that one:

1. I can replicate that in my local, for the picture tag with multiple sources we add link preload for each source and seems like we just need to get the first one.
2. I think this special character thing will be fixed by Opeyemy's PR
3. `fetchpriority='high'` is already added to the img tag inside picture.

MathieuLamiot commented:
@Miraeld @Mai-Saad About feedbacks 2 & 3, I tested commit 68f7ebeb780e728cddbd3ef62b1db7ecbabfd4bb (after running gulp build:js:all ) on [this page](http://mathieu.e2e.rocketlabsqa.ovh/lcp_6599_template2/).

The original markup is 

```
<picture  decoding="async" class="wp-image-8418">
  <source type="image/avif" srcset="https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1024x576.jpg.avif 1024w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-300x169.jpg.avif 300w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-768x432.jpg.avif 768w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1536x864.jpg.avif 1536w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1200x675.jpg.avif 1200w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-600x338.jpg.avif 600w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia.jpg.avif 1920w" sizes="(max-width: 1024px) 100vw, 1024px">
  <source type="image/webp" srcset="https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1024x576.jpg.webp 1024w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-300x169.jpg.webp 300w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-768x432.jpg.webp 768w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1536x864.jpg.webp 1536w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1200x675.jpg.webp 1200w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-600x338.jpg.webp 600w" sizes="(max-width: 1024px) 100vw, 1024px">
  <img  decoding="async" width="1024" height="576" src="https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1024x576.jpg" alt="" srcset="https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1024x576.jpg 1024w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-300x169.jpg 300w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-768x432.jpg 768w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1536x864.jpg 1536w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1200x675.jpg 1200w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-600x338.jpg 600w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia.jpg 1920w" sizes="(max-width: 1024px) 100vw, 1024px">
</picture>
```

The resulting preload is:

```
<link rel="preload" as="image" href="https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1024x576.jpg.avif 1024w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-300x169.jpg.avif 300w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-768x432.jpg.avif 768w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1536x864.jpg.avif 1536w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1200x675.jpg.avif 1200w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-600x338.jpg.avif 600w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia.jpg.avif 1920w" media="" fetchpriority="high">
<link rel="preload" as="image" href="https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1024x576.jpg.webp 1024w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-300x169.jpg.webp 300w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-768x432.jpg.webp 768w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1536x864.jpg.webp 1536w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1200x675.jpg.webp 1200w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-600x338.jpg.webp 600w" media="" fetchpriority="high">
```

So it looks like the special character thing is indeed fixed by updating to develop. But the preload is not an image source set, so the href is completely invalid here.

I don't get the fetch priority on the picture/img markup, only on the preload though. @wordpressfan Did you see it differently?

Let's discuss this after the daily.

wordpressfan commented:
![image](https://github.com/wp-media/wp-rocket/assets/146129302/2b8fdb4e-a1fd-4e4c-8042-eea6841f9aa8)

Yes I can see it

MathieuLamiot commented:
@wordpressfan As discussed orally, this is working on the picture tag you used, but not the one shared by @Mai-Saad (see [here](https://github.com/wp-media/wp-rocket/pull/6659#issuecomment-2147531800)).

I'll write a complete summary below.

MathieuLamiot commented:
### Summary of tests to perform to validate the PR, as of 2024-06-04 4:30PM

**Rocket E2E**
Update from develop (once #6693 is fixed & merged) and run Rocket-E2E.

**From #6669**

With this as LCP on a page:

```
<figure class="figure-core-image post-featured single-image title-wrap">
	<picture class="core-image">
		<source type="image/webp" media='(max-width: 500px)' srcset="https://variance.pl/wp-content/uploads/2024/05/Kwiatowy-Ksiezyc-400x600.webp">
		<source type="image/webp" media='(min-width: 501px) and (max-width: 768px)' srcset="https://variance.pl/wp-content/uploads/2024/05/Kwiatowy-Ksiezyc-768x513.webp">
		<img class="skip-lazy" src="https://variance.pl/wp-content/uploads/2024/05/Kwiatowy-Ksiezyc-1348x900.webp" alt="Kwiatowy księżyc" title="Kwiatowy księżyc">
	</picture>
</figure>
```
The preload tags must be valid, not have empty href and no duplicated min-width or max-width instructions.


**From #6660**
With this as LCP

```
<picture class="picture_test2"> 
			<source srcset="<?php echo content_url().'/rocket-test-data/images/paper.jpeg' ?>" media="(max-width: 400px)"> 
			<source  srcset="<?php echo content_url().'/rocket-test-data/images/lcp/testjpeg.jpeg' ?>" media="(max-width: 800px)"> 
			<img src="<?php echo content_url().'/rocket-test-data/images/lcp/testwebp.webp'?>"> 	
		</picture>
```
The preload tags must be valid, not have empty href and no duplicated min-width or max-width instructions.

**From https://github.com/wp-media/wp-rocket/issues/6638#issuecomment-2139333126**

This LCP:

```
<picture fetchpriority="high" decoding="async" class="wp-image-8418">
    <source type="image/avif" srcset="https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1024x576.jpg.avif 1024w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-300x169.jpg.avif 300w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-768x432.jpg.avif 768w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1536x864.jpg.avif 1536w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1200x675.jpg.avif 1200w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-600x338.jpg.avif 600w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia.jpg.avif 1920w" sizes="(max-width: 1024px) 100vw, 1024px">
    <source type="image/webp" srcset="https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1024x576.jpg.webp 1024w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-300x169.jpg.webp 300w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-768x432.jpg.webp 768w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1536x864.jpg.webp 1536w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1200x675.jpg.webp 1200w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-600x338.jpg.webp 600w" sizes="(max-width: 1024px) 100vw, 1024px">
    <img fetchpriority="high" decoding="async" width="1024" height="576" src="https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1024x576.jpg" alt="" srcset="https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1024x576.jpg 1024w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-300x169.jpg 300w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-768x432.jpg 768w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1536x864.jpg 1536w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1200x675.jpg 1200w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-600x338.jpg 600w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia.jpg 1920w" sizes="(max-width: 1024px) 100vw, 1024px">
</picture>
```

Must be preloaded as:

`<link rel="preload" as="image" imagesrcset="https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1024x576.jpg.avif 1024w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-300x169.jpg.avif 300w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-768x432.jpg.avif 768w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1536x864.jpg.avif 1536w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1200x675.jpg.avif 1200w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-600x338.jpg.avif 600w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia.jpg.avif 1920w" imagesizes="(max-width: 1024px) 100vw, 1024px" fetchpriority="high">`

This LCP:

```
	<picture decoding="async">
    <source type="image/avif" srcset="https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1024x576.jpg.avif">
    <source type="image/webp" srcset="https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1024x576.jpg.webp">
    <img decoding="async" src="https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1024x576.jpg">
</picture>
```

Must be preloaded as 
`<link rel="preload" as="image" href="https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1024x576.jpg.avif" fetchpriority="high">` 
and no others

Miraeld committed:
Fix how everything is handled for picture tag

Miraeld commented:
@wordpressfan @MathieuLamiot @Mai-Saad , 
In the latest commit I've just made, I've reworked the way we handle the front-end for preloading `<picture>` tags. 
Of course, it won't affect other tags as it's only used for `<picture>`. 

So, some tests and results based on last @MathieuLamiot comment:
**This markup**
``` html
<picture fetchpriority="high" decoding="async" class="wp-image-8418">
    <source type="image/avif" srcset="https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1024x576.jpg.avif 1024w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-300x169.jpg.avif 300w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-768x432.jpg.avif 768w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1536x864.jpg.avif 1536w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1200x675.jpg.avif 1200w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-600x338.jpg.avif 600w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia.jpg.avif 1920w" sizes="(max-width: 1024px) 100vw, 1024px">
    <source type="image/webp" srcset="https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1024x576.jpg.webp 1024w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-300x169.jpg.webp 300w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-768x432.jpg.webp 768w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1536x864.jpg.webp 1536w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1200x675.jpg.webp 1200w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-600x338.jpg.webp 600w" sizes="(max-width: 1024px) 100vw, 1024px">
    <img fetchpriority="high" decoding="async" width="1024" height="576" src="https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1024x576.jpg" alt="" srcset="https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1024x576.jpg 1024w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-300x169.jpg 300w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-768x432.jpg 768w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1536x864.jpg 1536w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1200x675.jpg 1200w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-600x338.jpg 600w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia.jpg 1920w" sizes="(max-width: 1024px) 100vw, 1024px">
</picture>
```
**will give the preloaded tag:**
``` html
<link rel="preload" as="image" imagesrcset="https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1024x576.jpg.avif 1024w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-300x169.jpg.avif 300w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-768x432.jpg.avif 768w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1536x864.jpg.avif 1536w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1200x675.jpg.avif 1200w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-600x338.jpg.avif 600w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia.jpg.avif 1920w" imagesizes="(max-width: 1024px) 100vw, 1024px" fetchpriority="high">
```

----

**Then this markup:**
``` html
	<picture decoding="async">
    <source type="image/avif" srcset="https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1024x576.jpg.avif">
    <source type="image/webp" srcset="https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1024x576.jpg.webp">
    <img decoding="async" src="https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1024x576.jpg">
</picture>
```

**is preloaded as:**
``` html
<link rel="preload" as="image" href="https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1024x576.jpg.avif" fetchpriority="high">
```

---

**This markup:**
``` html
<figure class="figure-core-image post-featured single-image title-wrap">
	<picture class="core-image">
		<source type="image/webp" media='(max-width: 500px)' srcset="https://variance.pl/wp-content/uploads/2024/05/Kwiatowy-Ksiezyc-400x600.webp">
		<source type="image/webp" media='(min-width: 501px) and (max-width: 768px)' srcset="https://variance.pl/wp-content/uploads/2024/05/Kwiatowy-Ksiezyc-768x513.webp">
		<img class="skip-lazy" src="https://variance.pl/wp-content/uploads/2024/05/Kwiatowy-Ksiezyc-1348x900.webp" alt="Kwiatowy księżyc" title="Kwiatowy księżyc">
	</picture>
</figure>
```

**is preloaded as:**
``` html
<link rel="preload" as="image" href="/wp-content/rocket-test-data/images/paper.jpeg" media="(max-width: 400px)" fetchpriority="high">
<link rel="preload" as="image" href="/wp-content/rocket-test-data/images/lcp/testjpeg.jpeg" media="(min-width: 400.1px) and (max-width: 800px)" fetchpriority="high">
<link rel="preload" as="image" href="http://localhost:8902/wp-content/rocket-test-data/images/lcp/testwebp.webp" media="(min-width: 800.1px)" fetchpriority="high">
```

-----

**Finally, this markup:**
``` php
<picture class="picture_test2"> 
			<source srcset="<?php echo content_url().'/rocket-test-data/images/paper.jpeg' ?>" media="(max-width: 400px)"> 
			<source  srcset="<?php echo content_url().'/rocket-test-data/images/lcp/testjpeg.jpeg' ?>" media="(max-width: 800px)"> 
			<img src="<?php echo content_url().'/rocket-test-data/images/lcp/testwebp.webp'?>"> 	
		</picture>
```

**is preloaded as:**
``` html
<link rel="preload" as="image" href="/wp-content/rocket-test-data/images/paper.jpeg" media="(max-width: 400px)" fetchpriority="high">
<link rel="preload" as="image" href="/wp-content/rocket-test-data/images/lcp/testjpeg.jpeg" media="(min-width: 400.1px) and (max-width: 800px)" fetchpriority="high">
<link rel="preload" as="image" href="/wp-content/rocket-test-data/images/lcp/testwebp.webp" media="(min-width: 800.1px)" fetchpriority="high">
```


----

I've updated the integration tests accordingly, and everything works fine. 

Also, the `fetchpriority="high"` is added to the `<img` tag within the `<picture>`

Miraeld committed:
Fix test 3 picture based on feedback markup

wordpressfan commented:
```
<figure class="figure-core-image post-featured single-image title-wrap">
	<picture class="core-image">
		<source type="image/webp" media='(max-width: 500px)' srcset="https://variance.pl/wp-content/uploads/2024/05/Kwiatowy-Ksiezyc-400x600.webp">
		<source type="image/webp" media='(min-width: 501px) and (max-width: 768px)' srcset="https://variance.pl/wp-content/uploads/2024/05/Kwiatowy-Ksiezyc-768x513.webp">
		<img class="skip-lazy" src="https://variance.pl/wp-content/uploads/2024/05/Kwiatowy-Ksiezyc-1348x900.webp" alt="Kwiatowy księżyc" title="Kwiatowy księżyc">
	</picture>
</figure>
```

Preloaded as:

```
<link rel="preload" as="image" href="https://variance.pl/wp-content/uploads/2024/05/Kwiatowy-Ksiezyc-400x600.webp" media="(max-width: 500px)" fetchpriority="high">
<link rel="preload" as="image" href="https://variance.pl/wp-content/uploads/2024/05/Kwiatowy-Ksiezyc-1348x900.webp" media="(min-width: 500.1px)" fetchpriority="high">
```

-----------------------------------------

```
<picture class="picture_test2">
    <source srcset="https://new.rocketlabsqa.ovh/wp-content/rocket-test-data/images/paper.jpeg" media="(max-width: 400px)">
    <source  srcset="https://new.rocketlabsqa.ovh/wp-content/rocket-test-data/images/lcp/testjpeg.jpeg" media="(max-width: 800px)">
    <img src="https://new.rocketlabsqa.ovh/wp-content/rocket-test-data/images/lcp/testwebp.webp">
</picture>
```

Preloaded as:

```
<link rel="preload" as="image" href="https://new.rocketlabsqa.ovh/wp-content/rocket-test-data/images/paper.jpeg" media="(max-width: 400px)" fetchpriority="high">
<link rel="preload" as="image" href="https://new.rocketlabsqa.ovh/wp-content/rocket-test-data/images/lcp/testjpeg.jpeg" media="(min-width: 400.1px) and (max-width: 800px)" fetchpriority="high">
<link rel="preload" as="image" href="https://new.rocketlabsqa.ovh/wp-content/rocket-test-data/images/lcp/testwebp.webp" media="(min-width: 800.1px)" fetchpriority="high">
```

-----------------------------------------

```
<picture fetchpriority="high" decoding="async" class="wp-image-8418">
    <source type="image/avif" srcset="https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1024x576.jpg.avif 1024w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-300x169.jpg.avif 300w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-768x432.jpg.avif 768w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1536x864.jpg.avif 1536w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1200x675.jpg.avif 1200w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-600x338.jpg.avif 600w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia.jpg.avif 1920w" sizes="(max-width: 1024px) 100vw, 1024px">
    <source type="image/webp" srcset="https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1024x576.jpg.webp 1024w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-300x169.jpg.webp 300w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-768x432.jpg.webp 768w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1536x864.jpg.webp 1536w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1200x675.jpg.webp 1200w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-600x338.jpg.webp 600w" sizes="(max-width: 1024px) 100vw, 1024px">
    <img fetchpriority="high" decoding="async" width="1024" height="576" src="https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1024x576.jpg" alt="" srcset="https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1024x576.jpg 1024w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-300x169.jpg 300w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-768x432.jpg 768w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1536x864.jpg 1536w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1200x675.jpg 1200w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-600x338.jpg 600w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia.jpg 1920w" sizes="(max-width: 1024px) 100vw, 1024px">
</picture>
```

Preloaded as:

```
<link rel="preload" as="image" imagesrcset="https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1024x576.jpg.avif 1024w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-300x169.jpg.avif 300w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-768x432.jpg.avif 768w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1536x864.jpg.avif 1536w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1200x675.jpg.avif 1200w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-600x338.jpg.avif 600w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia.jpg.avif 1920w" imagesizes="(max-width: 1024px) 100vw, 1024px" fetchpriority="high">
```

-----------------------------------------

```
<picture decoding="async">
    <source type="image/avif" srcset="https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1024x576.jpg.avif">
    <source type="image/webp" srcset="https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1024x576.jpg.webp">
    <img decoding="async" src="https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1024x576.jpg">
</picture>
```

Preloaded as:

```
<link rel="preload" as="image" href="[https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1024x576.jpg.avif](https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-%E2%80%94-kopia-1024x576.jpg.avif)" fetchpriority="high">
```

I'm trying to compare those results with the results mentioned by Mathieu's last comment.

MathieuLamiot commented:
@Miraeld I see you moved this to Ready for Review.
But it is not up-to-date with develop and there are no E2E results shared for this PR updated on develop. Can you make sure to provide all this?

Miraeld committed:
Merge branch 'develop' into enhancement/6638-handle-picture-tag-in-lcp-preload

Miraeld commented:
The codebase is ready for review, 
I'm trying to run e2e, but can't get it working (the whole e2e app), non related to this PR. 

And it's just been updated with develop latest changes.

Mai-Saad commented:
@wordpressfan Thanks for sharing markups [above](https://github.com/wp-media/wp-rocket/pull/6659#issuecomment-2149633674).

- For the 1st markup
```

	<picture class="core-image">
		<source type="image/webp" media='(max-width: 500px)' srcset="https://variance.pl/wp-content/uploads/2024/05/Kwiatowy-Ksiezyc-400x600.webp">
		<source type="image/webp" media='(min-width: 501px) and (max-width: 768px)' srcset="https://variance.pl/wp-content/uploads/2024/05/Kwiatowy-Ksiezyc-768x513.webp">
		<img class="skip-lazy" src="https://variance.pl/wp-content/uploads/2024/05/Kwiatowy-Ksiezyc-1348x900.webp" alt="Kwiatowy księżyc" title="Kwiatowy księżyc">
	</picture>

```
**Update: it's 2nd source missing and thus min-width is not correct** thanks to @Miraeld 

 it's missing to preload `<img> ` + and max-width of 2nd image => clear on this page https://new.rocketlabsqa.ovh/lcp_picture_media_type_mixed/. @wp-media/productrocket it contains both min-width and max-width, **please confirm expected here.** (now no empty href nor duplicated min-width but we didn't preload img tag and we removed and max-width from 2nd image)
Expected markup as per original issue https://github.com/wp-media/wp-rocket/issues/6669:
```
<link rel="preload" as="image" href="https://variance.pl/wp-content/uploads/2024/05/Kwiatowy-Ksiezyc-400x600.webp" media="(max-width: 500px)" fetchpriority="high">
<link rel="preload" as="image" href="https://variance.pl/wp-content/uploads/2024/05/Kwiatowy-Ksiezyc-768x513.webp" media="(min-width: 501px) and (max-width: 768px)" fetchpriority="high">
<link rel="preload" as="image" href="https://variance.pl/wp-content/uploads/2024/05/Kwiatowy-Ksiezyc-1348x900.webp" alt="Kwiatowy księżyc" media="(min-width: 768.1px)" fetchpriority="high">
```
- 2nd markup based on https://github.com/wp-media/wp-rocket/issues/6530 is :+1:  (test page: https://new.rocketlabsqa.ovh/lcp_picture_template2/)
- 3rd markup, based on https://github.com/wp-media/wp-rocket/issues/6638#issuecomment-2139333126 is :+1:  (test page: https://new.rocketlabsqa.ovh/lcp_picture_responsive_type_template/) 
- 4th markup, based on 2nd point here https://github.com/wp-media/wp-rocket/issues/6638#issuecomment-2139333126 is :+1:  (test page https://new.rocketlabsqa.ovh/lcp_picture_type/)

MathieuLamiot commented:
E2E results: https://wp-media.slack.com/archives/CUT7FLHF1/p1717653110913929?thread_ts=1717624428.972699&cid=CUT7FLHF1
Moving this to Ready for Review. And pending Product feedback.

Miraeld commented:
I was checking on `develop`, 
while using the HTML markup:
``` html
	<picture class="core-image">
		<source type="image/webp" media='(max-width: 500px)' srcset="https://variance.pl/wp-content/uploads/2024/05/Kwiatowy-Ksiezyc-400x600.webp">
		<source type="image/webp" media='(min-width: 501px) and (max-width: 768px)' srcset="https://variance.pl/wp-content/uploads/2024/05/Kwiatowy-Ksiezyc-768x513.webp">
		<img class="skip-lazy" src="https://variance.pl/wp-content/uploads/2024/05/Kwiatowy-Ksiezyc-1348x900.webp" alt="Kwiatowy księżyc" title="Kwiatowy księżyc">
	</picture>

```
@piotrbak for everything related to the only markup @Mai-Saad mentioned was wrong, on develop we are getting the following preload tag:
``` html
<link rel="preload" as="image" href="https://variance.pl/wp-content/uploads/2024/05/Kwiatowy-Ksiezyc-1348x900.webp" fetchpriority="high">
```


What we are getting on this PR:
``` html
<link rel="preload" as="image" href="https://variance.pl/wp-content/uploads/2024/05/Kwiatowy-Ksiezyc-400x600.webp" media="(max-width: 500px)" fetchpriority="high">
<link rel="preload" as="image" href="https://variance.pl/wp-content/uploads/2024/05/Kwiatowy-Ksiezyc-1348x900.webp" media="(min-width: 500.1px)" fetchpriority="high">
```

Not sure if this is a regression or not what we are getting with this PR. 
What is the actual expected behavior for this ? Is @Mai-Saad correct in her latest comment :
```  html
<link rel="preload" as="image" href="https://variance.pl/wp-content/uploads/2024/05/Kwiatowy-Ksiezyc-400x600.webp" media="(max-width: 500px)" fetchpriority="high">
<link rel="preload" as="image" href="https://variance.pl/wp-content/uploads/2024/05/Kwiatowy-Ksiezyc-768x513.webp" media="(min-width: 501px) and (max-width: 768px)" fetchpriority="high">
<link rel="preload" as="image" href="https://variance.pl/wp-content/uploads/2024/05/Kwiatowy-Ksiezyc-1348x900.webp" alt="Kwiatowy księżyc" media="(min-width: 768.1px)" fetchpriority="high">
```

I was thinking to move this case out of this PR, so everything else can move forward while this one is being worked on. 
What do you think ?


EDIT: 
After looking into it a bit, it isn't the `<img` tag that isn't preloaded, but the second `<source>` tag. 
This is due to this bail-out [here](https://github.com/wp-media/wp-rocket/blob/0aa0cf18afff96d3bf265aae3ca55fcf6abf91a7/inc/Engine/Media/AboveTheFold/Frontend/Controller.php#L492-L494), which was introduce to tackle another case of the picture tag.

Miraeld committed:
Fix last picture error when we are preloading lcp

Miraeld committed:
Merge branch 'develop' into enhancement/6638-handle-picture-tag-in-lcp-preload

Miraeld committed:
Update lcp-beacon

Miraeld commented:
@Mai-Saad , 
As per discussion [here](https://wp-media.slack.com/archives/CUT7FLHF1/p1717683402625259), we can move forward with this PR, while fixing the remaining markup in another issue [here](https://github.com/wp-media/wp-rocket/issues/6707)

Miraeld committed:
Merge branch 'enhancement/6638-handle-picture-tag-in-lcp-preload' into fix/picture-markup-type-bailout

MathieuLamiot committed:
Add back src for picture and img-srcset

MathieuLamiot committed:
Merge commit '4cd5316a7b07ad8ee07e207760365dbc07553ed8' into fix/picture-src-regression

MathieuLamiot committed:
Merge with picture fix PR and gulp

MathieuLamiot committed:
fixl inter

Miraeld commented:
@MathieuLamiot , @Mai-Saad , 
This PR Now includes changes from https://github.com/wp-media/wp-rocket/pull/6708, & https://github.com/wp-media/wp-rocket/pull/6709. 
Which means the picture tag should work properly now, on every aspect + the `src` regression is fixed.

Mai-Saad commented:
@Miraeld Thanks for the update.
Tested templates:
- https://new.rocketlabsqa.ovh/lcp_picture_media_type_mixed/ :+1: 
- https://new.rocketlabsqa.ovh/lcp_picture_relative/ :+1: 
- https://new.rocketlabsqa.ovh/lcp_picture_responsive_type/ :+1: 
- https://new.rocketlabsqa.ovh/lcp_picture_template2/ :+1: 
- https://new.rocketlabsqa.ovh/lcp_picture_type/ :+1: 
- https://new.rocketlabsqa.ovh/lcp_bg_multimage_template/ :+1: 
- https://new.rocketlabsqa.ovh/lcp_responsive_image_template/ :+1:  (noted here and already on trunk, if img not having sizes we will have empty imagesizes in preload which will cause markup error, however, already on nowprocket, not having size to responsive image was causing error) ==> we can update the test template later to ensure original markup is valid

- General note: markup with only `media='min-width'` will only preload srcset now (accepted as edge case for now) https://new.rocketlabsqa.ovh/lcp_picture_template/

Possible breaking markups (not sure if it's common, needs further investigation and expected o/p from @wp-media/productrocket if we will handle any in new GH maybe):

1. test page https://new.rocketlabsqa.ovh/lcp_picture_responsive_type_mixed/
 ```
      <picture   class="wp-image-8418">
    <source Type="image/avif" Media = '(max-width: 500px)' Srcset="https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1024x576.jpg.avif 1024w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-300x169.jpg.avif 300w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-768x432.jpg.avif 768w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1536x864.jpg.avif 1536w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1200x675.jpg.avif 1200w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-600x338.jpg.avif 600w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia.jpg.avif 1920w" sizes="(max-width: 1024px) 100vw, 1024px">
    <source type = "image/webp" Media='(max-width: 800px)' SRCSET="https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1024x576.jpg.webp 1024w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-300x169.jpg.webp 300w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-768x432.jpg.webp 768w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1536x864.jpg.webp 1536w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1200x675.jpg.webp 1200w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-600x338.jpg.webp 600w" sizes="(max-width: 1024px) 100vw, 1024px">
    <img  width="1024" height="576" Src="https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1024x576.jpg" alt="" srcset="https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1024x576.jpg 1024w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-300x169.jpg 300w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-768x432.jpg 768w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1536x864.jpg 1536w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1200x675.jpg 1200w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-600x338.jpg 600w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia.jpg 1920w" sizes="(max-width: 1024px) 100vw, 1024px">
</picture>
```
will be preloaded as
```<link rel="preload" as="image" imagesrcset="https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1024x576.jpg.avif 1024w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-300x169.jpg.avif 300w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-768x432.jpg.avif 768w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1536x864.jpg.avif 1536w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-1200x675.jpg.avif 1200w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia-600x338.jpg.avif 600w, https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-—-kopia.jpg.avif 1920w" media="(max-width: 500px)" imagesizes="(max-width: 1024px) 100vw, 1024px" fetchpriority="high">
<link rel="preload" as="image" href="https://imagify.rocketlabsqa.ovh/wp-content/uploads/2024/05/home-new-bg-free-img-%E2%80%94-kopia-1024x576.jpg" media="(min-width: 500.1px)" fetchpriority="high">
```

2- using max-height instead of max-width https://new.rocketlabsqa.ovh/lcp_picture_template2_maxheight/
```
<picture class="picture_test2"> 
			<source srcset="<?php echo content_url().'/rocket-test-data/images/paper.jpeg' ?>" media="(max-height: 400px)"> 
			<source  srcset="<?php echo content_url().'/rocket-test-data/images/lcp/testjpeg.jpeg' ?>" media="(max-height: 800px)"> 
			<img src="<?php echo content_url().'/rocket-test-data/images/lcp/testwebp.webp'?>"> 	
		</picture>
```
will be

```
<link rel="preload" as="image" href="https://new.rocketlabsqa.ovh/wp-content/rocket-test-data/images/paper.jpeg" media="(max-height: 400px)" fetchpriority="high">
<link rel="preload" as="image" href="https://new.rocketlabsqa.ovh/wp-content/rocket-test-data/images/lcp/testjpeg.jpeg" media="(max-height: 800px)" fetchpriority="high">
```

- e2e results https://wp-media.slack.com/archives/CUT7FLHF1/p1717836779879489?thread_ts=1717754667.187049&cid=CUT7FLHF1
Recommend rerun e2e and if all good + we won't handle markups above here then we can merge

MathieuLamiot commented:
Re-ran E2E to confirm: all good (details [here](https://wp-media.slack.com/archives/CUT7FLHF1/p1717866632267799?thread_ts=1717754667.187049&cid=CUT7FLHF1)).

Only waiting for @piotrbak @DahmaniAdame confirmation on the 2 points from @Mai-Saad above.

piotrbak commented:
@MathieuLamiot @Mai-Saad We're good with those
